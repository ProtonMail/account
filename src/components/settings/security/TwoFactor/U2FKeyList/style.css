.listHeader {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
}

.listElement {
    display: flex;
    justify-content: space-between;
    border: 1px solid black;
}

.list {
    display: flex;
    flex-direction: column;
}
