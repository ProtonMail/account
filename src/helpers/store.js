import { createStore } from 'unistore/full/preact';
import devtools from 'unistore/devtools';

import authActions from '../actions/authentication';
import settingsActions from '../actions/settings';
import scopeActions from '../actions/scope';

export const initialState = {
    auth: {
        isLoggedIn: false,
        user: {},
        step: 'login',
        twoFactorResponse: {}
    },
    scope: {
        creds: null,
        response: null
    },
    settings: {
        addU2FKey: {},
        reset2FARecoveryCodes: {}
    }
};
const store = createStore(initialState);


export const actions = (store) => ({
    ...authActions(store),
    ...settingsActions(store),
    ...scopeActions(store)
});

export default store;
