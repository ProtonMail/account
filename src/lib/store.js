import { createStore } from 'unistore/full/preact';

import authActions from '../actions/authentication';
import settingsActions from '../actions/settings';

const store = createStore({
    auth: {
        isLoggedIn: false,
        user: {},
        step: 'login',
        twoFactorResponse: {}
    },
    settings: {
        addU2FKey: {},
        reset2FARecoveryCodes: {}
    }
});

export const actions = (store) => ({
    ...authActions(store),
    ...settingsActions(store)
});

export default store;
