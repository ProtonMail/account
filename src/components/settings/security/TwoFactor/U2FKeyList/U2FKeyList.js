import { h, Component } from 'preact';
import { connect } from 'unistore/full/preact';

import settingsActions from '../../../../../actions/settings';
import ConfirmModal from '../../../../ui/ConfirmModal';
import TextButton from '../../../../ui/TextButton';

import style from './style.css';

export class U2FKeyList extends Component {
    /**
     * renders an U2F Key.
     * @param {Object} u2fKey - the U2F Key to render.
     * @param {Int} u2fKey.Compromised - whether the key is compromised or not.
     * @param {String} u2fKey.KeyHandle - The key handle of the current key.
     * @param {String} u2fKey.Label - The label of the current key.
     * @return {Component}
     */
    renderU2FKey(u2fKey) {
        const headerClasses = [style.listElementHeader];
        if (u2fKey.Compromised) {
            headerClasses.push(style.listElementHeaderCompromised);
        }

        return (
            <li key={u2fKey.KeyHandle} className={style.listElement}>
                <div className={headerClasses.join(' ')}>{u2fKey.Label}</div>
                {!!u2fKey.Compromised && <div class="badge badge-danger">Compromised</div>}
                <TextButton
                    onClick={() =>
                        this.setState({
                            confirmDeleteModal: u2fKey
                        })
                    }
                >
                    Delete
                </TextButton>
            </li>
        );
    }

    render() {
        const confirmDeleteModal = this.state.confirmDeleteModal;

        const closeModal = () => {
            this.setState({ confirmDeleteModal: '' });
        };
        return (
            <ul className={style.list}>
                <ConfirmModal
                    title="Confirm Security Key Deletion"
                    scope="password"
                    isOpen={!!confirmDeleteModal}
                    onAfterClose={closeModal}
                    onConfirm={() => this.props.deleteU2FKeyAction(confirmDeleteModal)}
                    onCancel={() => {}}
                >
                    <div>
                        Are you sure you want to delete the key{' '}
                        <span>{confirmDeleteModal ? confirmDeleteModal.Label : ''}</span>?
                    </div>
                </ConfirmModal>
                {this.props.U2FKeys.map((u2fKey) => this.renderU2FKey(u2fKey))}
            </ul>
        );
    }
}

export default connect(
    'settings',
    settingsActions
)(U2FKeyList);
