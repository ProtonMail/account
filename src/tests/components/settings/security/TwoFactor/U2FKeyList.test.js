import { U2FKeyList } from '../../../../../components/settings/security/TwoFactor/U2FKeyList/U2FKeyList';

import render from 'preact-render-to-string';
import { deep } from 'preact-render-spy';

jest.mock('../../../../../actions/settings');


import { deepProvider, renderProvided, shallowProvider } from '../../../../testsHelpers/storeTools';
import store from '../../../../../helpers/store';

describe('U2FKeyList test', () => {
    test('with no key', () => {
        expect(render(<U2FKeyList U2FKeys={[]}/>)).toMatchSnapshot();
    });

    test('with keys', () => {
        const keys = [
            { KeyHandle: '1', Compromised: false, Label: 'First key' },
            { KeyHandle: '2', Compromised: false, Label: 'Second key' },
            { KeyHandle: '3', Compromised: true, Label: 'Compromised key' },
            { KeyHandle: '4', Compromised: false, Label: 'Last key' }
        ];
        expect(render(<U2FKeyList U2FKeys={keys}/>)).toMatchSnapshot();
    });

    test('opening modal', () => {
        const keys = [
            { KeyHandle: '1', Compromised: false, Label: 'First key' },
            { KeyHandle: '2', Compromised: false, Label: 'Second key' },
            { KeyHandle: '3', Compromised: true, Label: 'Compromised key' },
            { KeyHandle: '4', Compromised: false, Label: 'Last key' }
        ];
        const context = deep(<U2FKeyList U2FKeys={keys}/>);
        expect(context.state('confirmDeleteModal')).toBeFalsy();
        context.find('button').first().simulate('click');
        expect(context).toMatchSnapshot();
        expect(context.state('confirmDeleteModal')).toMatchObject(expect.objectContaining({
            KeyHandle: expect.any(String),
            Label: expect.any(String)
        }));
    });
});
