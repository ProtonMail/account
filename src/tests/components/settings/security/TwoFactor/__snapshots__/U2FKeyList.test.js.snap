// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`U2FKeyList test opening modal 1`] = `
preact-render-spy (1 nodes)
-------
<ul class="list">
  <li class="listElement">
    <div class="listElementHeader">First key</div>
    <button onClick={[Function onClick]}>Delete</button>
  </li>
  <li class="listElement">
    <div class="listElementHeader">Second key</div>
    <button onClick={[Function onClick]}>Delete</button>
  </li>
  <li class="listElement">
    <div class="listElementHeader listElementHeaderCompromised">Compromised key</div>
    <div class="badge badge-danger">Compromised</div>
    <button onClick={[Function onClick]}>Delete</button>
  </li>
  <li class="listElement">
    <div class="listElementHeader">Last key</div>
    <button onClick={[Function onClick]}>Delete</button>
  </li>
</ul>

`;

exports[`U2FKeyList test with keys 1`] = `
<ul class="list">
  <li class="listElement">
    <div class="listElementHeader">First key</div>
    <button>Delete</button>
  </li>
  <li class="listElement">
    <div class="listElementHeader">Second key</div>
    <button>Delete</button>
  </li>
  <li class="listElement">
    <div class="listElementHeader listElementHeaderCompromised">Compromised key</div>
    <div class="badge badge-danger">Compromised</div>
    <button>Delete</button>
  </li>
  <li class="listElement">
    <div class="listElementHeader">Last key</div>
    <button>Delete</button>
  </li>
</ul>
`;

exports[`U2FKeyList test with no key 1`] = `<ul class="list"></ul>`;
